drivers/timer/nrf_rtc_timer.c

```diff
@@ -117,21 +130,53 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
 	ticks = (ticks == K_FOREVER) ? MAX_TICKS : ticks;
 	ticks = max(min(ticks - 1, (s32_t)MAX_TICKS), 0);
 
-	k_spinlock_key_t key = k_spin_lock(&lock);
-	u32_t cyc, t = counter();
-
-	/* Round up to next tick boundary */
-	cyc = ticks * CYC_PER_TICK + counter_sub(t, last_count);
-	cyc += (CYC_PER_TICK - 1);
-	cyc = (cyc / CYC_PER_TICK) * CYC_PER_TICK;
-	cyc += last_count;
-
-	if (counter_sub(cyc, t) < MIN_DELAY) {
-		cyc += CYC_PER_TICK;
+	/*
+	 * Get the requested delay in tick-aligned cycles.  Increase
+	 * by one tick to round up so we don't timeout early due to
+	 * cycles elapsed since the last tick.  Cap at the maximum
+	 * tick-aligned delta.
+	 */
+	u32_t cyc = min((1 + ticks) * CYC_PER_TICK, MAX_DELAY);
+
+	u32_t key = irq_lock();
+	u32_t d = counter_sub(counter(), last_count);
+
+	/*
+	 * We've already accounted for anything less than a full tick,
+	 * and assumed we meet the minimum delay for the tick.  If
+	 * that's not true, we have to adjust, which may involve a
+	 * rare and expensive integer division.
+	 */
+	if (d >= (CYC_PER_TICK - MIN_DELAY)) {
+		if (d >= CYC_PER_TICK) {
+			/*
+			 * We're late by at least one tick.  Adjust
+			 * the compare offset for the missed ones, and
+			 * reduce d to be the portion since the last
+			 * (unseen) tick.
+			 */
+			u32_t missed_ticks = d / CYC_PER_TICK;
+			u32_t missed_cycles = missed_ticks * CYC_PER_TICK;
+			cyc += missed_cycles;
+			d -= missed_cycles;
+		}
+		if (d >= (CYC_PER_TICK - MIN_DELAY)) {
+			/*
+			 * We're (now) within the tick, but too close
+			 * to meet the minimum delay required to
+			 * guarantee compare firing.  Step up to the
+			 * next tick.
+			 */
+			cyc += CYC_PER_TICK;
+		}
+		if (cyc > MAX_DELAY) {
+			/* Don't adjust beyond the counter range. */
+			cyc = MAX_DELAY;
+		}
 	}
+	set_comparator(last_count + cyc);
```

Comment:
We should fire on tick boundary. Could we just keep adding CYC_PER_TICK to last_count until we met MIN_DELAY?

Reply:
We do fire on a tick boundary.  `last_count` is tick-aligned, and `cyc` is a multiple of clocks-per-tick.

Reply:
I have no doubt about that. I am just curious, if we could replace a series of if's above by the simpler code.

Reply:
I really don't think so.  The maximum time required for this code can be measured.  If we put a loop in, we have to worry about it iterating so many times that counter will have incremented more than twice and `MIN_DELAY` is no longer sufficient.

There are multiple conditions:
* We can be executing this within the tick recorded in `last_count` or in a later tick.  The first is the common case; the second requires expensive adds that are unrelated to `MIN_DELAY`, handled in the first inner condition.
* We can be more than `MIN_DELAY`  before the tick we're aiming for, or less.  The first is the common case; the second requires adds related to `MIN_DELAY`.  This can apply whether or not we needed to handle the first condition, and so is handled in the second condition.
* We can end up with cycles that exceed the counter representation.  This would break tick alignment unless ticks-per-sec evenly divides 32768.  We only have to worry about that if we did any adds, but in that case we can use a pre-computed maximum.

I don't think there's any way to simplify that.

Reply:
Ok.

Reply:
We can, I think, change `>=` to `>` in the comparisons related to `MIN_DELAY`.  Lemme try that in the stress-test application.

Reply:
I've adjusted the conditionals to be more strict, seems to work.  I'll let it keep running.